### PR TITLE
[Merged by Bors] - feat(Algebra/Category/Grp/Limits): the forgetful functor on the category of groups creates all limits

### PR DESCRIPTION
**Pull request by smorel394**
The forgetful functor on the category of groups (commutative or not) creates all limits.

As we already know that the forgetful functor from groups to monoids creates all limits (once we remove the unnecessary size condition in `Grp.Forget₂.createsLimit`), this just follows from the similar fact for monoid categories, which is PR #19965.

- [x] depends on: #19965 

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Category/Grp/Limits.lean [hunk 1]
```diff
@@ -86,7 +84,13 @@ noncomputable instance Forget₂.createsLimit :
   letI : (forget₂ Grp.{u} MonCat.{u}).ReflectsIsomorphisms :=
     CategoryTheory.reflectsIsomorphisms_forget₂ _ _
   createsLimitOfReflectsIso (K := F) (F := (forget₂ Grp.{u} MonCat.{u}))
-    fun c' t =>
+    fun c' t => by
+      have : Small.{u} (Functor.sections ((F ⋙ forget₂ Grp MonCat) ⋙ forget MonCat)) := by
+        have : HasLimit (F ⋙ forget₂ Grp MonCat) := HasLimit.mk {cone := c', isLimit := t}
```
Comment (joelriou) [inline Mathlib/Algebra/Category/Grp/Limits.lean:89]:
```suggestion
        have : HasLimit (F ⋙ forget₂ Grp MonCat) := ⟨_, t⟩
```

File: Mathlib/Algebra/Category/Grp/Limits.lean [hunk 2]
```diff
@@ -86,7 +84,13 @@ noncomputable instance Forget₂.createsLimit :
...
+        apply Concrete.small_sections_of_hasLimit (F ⋙ forget₂ Grp MonCat)
+      have : Small.{u} (Functor.sections (F ⋙ forget Grp)) := inferInstanceAs <| Small.{u}
+        (Functor.sections ((F ⋙ forget₂ Grp MonCat) ⋙ forget MonCat))
+      exact
```
Comment (joelriou) [inline Mathlib/Algebra/Category/Grp/Limits.lean:93]: `exact` is unnecessary here if you remove the `by` at the `fun c' t => by` line.
```suggestion
```
The `{ liftedCone := ... }` block should be indented by two more spaces.

File: Mathlib/Algebra/Category/Grp/Limits.lean [hunk 3]
```diff
@@ -193,6 +197,33 @@ instance forget_preservesLimitsOfSize :
 instance forget_preservesLimits : PreservesLimits (forget Grp.{u}) :=
   Grp.forget_preservesLimitsOfSize.{u, u}
 
+@[to_additive]
+noncomputable instance forget_createsLimit :
+    CreatesLimit F (forget Grp.{u}) := by
+  have : CreatesLimit F (forget₂ Grp.{u} MonCat.{u} ⋙ forget MonCat.{u}) := by
+    exact compCreatesLimit (forget₂ Grp.{u} MonCat.{u}) (forget MonCat.{u})
+  set e : forget₂ Grp.{u} MonCat.{u} ⋙ forget MonCat.{u} ≅ forget Grp.{u} :=
+    NatIso.ofComponents (fun _ ↦ Iso.refl _) (fun _ ↦ rfl)
```
Comment (joelriou) [inline Mathlib/Algebra/Category/Grp/Limits.lean:206]:
```suggestion
  set e : forget₂ Grp.{u} MonCat.{u} ⋙ forget MonCat.{u} ≅ forget Grp.{u} := Iso.refl _
```

File: Mathlib/Algebra/Category/Grp/Limits.lean [hunk 4]
```diff
@@ -193,6 +197,33 @@ instance forget_preservesLimitsOfSize :
 instance forget_preservesLimits : PreservesLimits (forget Grp.{u}) :=
   Grp.forget_preservesLimitsOfSize.{u, u}
 
+@[to_additive]
+noncomputable instance forget_createsLimit :
+    CreatesLimit F (forget Grp.{u}) := by
+  have : CreatesLimit F (forget₂ Grp.{u} MonCat.{u} ⋙ forget MonCat.{u}) := by
+    exact compCreatesLimit (forget₂ Grp.{u} MonCat.{u}) (forget MonCat.{u})
```
Comment (joelriou) [inline Mathlib/Algebra/Category/Grp/Limits.lean:204]: This is inferred automatically:
```suggestion
```

File: Mathlib/Algebra/Category/Grp/Limits.lean [hunk 5]
```diff
@@ -193,6 +197,33 @@ instance forget_preservesLimitsOfSize :
...
+  set e : forget₂ Grp.{u} MonCat.{u} ⋙ forget MonCat.{u} ≅ forget Grp.{u} :=
+    NatIso.ofComponents (fun _ ↦ Iso.refl _) (fun _ ↦ rfl)
+  exact createsLimitOfNatIso e
+
+@[to_additive]
+noncomputable instance forget_createsLimitsOfShape :
+    CreatesLimitsOfShape J (forget Grp.{u}) where
+      CreatesLimit := inferInstance
```
Comment (joelriou) [inline Mathlib/Algebra/Category/Grp/Limits.lean:212]:
```suggestion
  CreatesLimit := inferInstance
```

File: Mathlib/Algebra/Category/Grp/Limits.lean [hunk 6]
```diff
@@ -193,6 +197,33 @@ instance forget_preservesLimitsOfSize :
...
+
+/-- The forgetful functor from groups to types creates all limits.
+-/
+@[to_additive
+  "The forgetful functor from additive groups to types creates all limits.",
+  to_additive_relevant_arg 2]
+noncomputable instance forget_createsLimitsOfSize :
+    CreatesLimitsOfSize.{w,v} (forget Grp.{u}) where
+      CreatesLimitsOfShape := inferInstance
```
Comment (joelriou) [inline Mathlib/Algebra/Category/Grp/Limits.lean:221]:
```suggestion
  CreatesLimitsOfShape := inferInstance
```

File: Mathlib/Algebra/Category/Grp/Limits.lean [hunk 7]
```diff
@@ -398,6 +429,31 @@ noncomputable instance _root_.AddCommGrp.forget_preservesLimits :
 noncomputable instance forget_preservesLimits : PreservesLimits (forget CommGrp.{u}) :=
   CommGrp.forget_preservesLimitsOfSize.{u, u}
 
+@[to_additive]
+noncomputable instance forget_createsLimit :
+    CreatesLimit F (forget CommGrp.{u}) := by
+  set e : forget₂ CommGrp.{u} Grp.{u} ⋙ forget Grp.{u} ≅ forget CommGrp.{u} :=
+    NatIso.ofComponents (fun _ ↦ Iso.refl _) (fun _ ↦ rfl)
```
Comment (joelriou) [inline Mathlib/Algebra/Category/Grp/Limits.lean:436]:
```suggestion
  set e : forget₂ CommGrp.{u} Grp.{u} ⋙ forget Grp.{u} ≅ forget CommGrp.{u} := Iso.refl _
```

File: Mathlib/Algebra/Category/Grp/Limits.lean [hunk 8]
```diff
@@ -398,6 +429,31 @@ noncomputable instance _root_.AddCommGrp.forget_preservesLimits :
...
+  exact createsLimitOfNatIso e
+
+@[to_additive]
+noncomputable instance forget_cratesLimitsOfShape (J : Type v) [Category.{w} J] :
+    CreatesLimitsOfShape J (forget CommGrp.{u}) where
+      CreatesLimit := inferInstance
```
Comment (joelriou) [inline Mathlib/Algebra/Category/Grp/Limits.lean:442]:
```suggestion
  CreatesLimit := inferInstance
```

File: Mathlib/Algebra/Category/Grp/Limits.lean [hunk 9]
```diff
@@ -398,6 +429,31 @@ noncomputable instance _root_.AddCommGrp.forget_preservesLimits :
...
+
+/-- The forgetful functor from commutative groups to types creates all limits.
+-/
+@[to_additive
+  "The forgetful functor from additive commutative groups to types creates all limits.",
+  to_additive_relevant_arg 2]
+noncomputable instance forget_createsLimitsOfSize :
+    CreatesLimitsOfSize.{w, v} (forget CommGrp.{u}) where
+      CreatesLimitsOfShape := inferInstance
```
Comment (joelriou) [inline Mathlib/Algebra/Category/Grp/Limits.lean:451]:
```suggestion
  CreatesLimitsOfShape := inferInstance
```

File: Mathlib/Algebra/Category/Grp/Limits.lean [hunk 10]
```diff
@@ -398,6 +429,31 @@ noncomputable instance _root_.AddCommGrp.forget_preservesLimits :
...
 
+@[to_additive]
+noncomputable instance forget_createsLimit :
+    CreatesLimit F (forget CommGrp.{u}) := by
+  set e : forget₂ CommGrp.{u} Grp.{u} ⋙ forget Grp.{u} ≅ forget CommGrp.{u} :=
+    NatIso.ofComponents (fun _ ↦ Iso.refl _) (fun _ ↦ rfl)
+  exact createsLimitOfNatIso e
+
+@[to_additive]
+noncomputable instance forget_cratesLimitsOfShape (J : Type v) [Category.{w} J] :
```
Comment (joelriou) [inline Mathlib/Algebra/Category/Grp/Limits.lean:440]:
```suggestion
noncomputable instance forget_createsLimitsOfShape (J : Type v) [Category.{w} J] :
```

File: Mathlib/Algebra/Category/Grp/Limits.lean [hunk 11]
```diff
@@ -398,6 +429,31 @@ noncomputable instance _root_.AddCommGrp.forget_preservesLimits :
...
+  "The forgetful functor from additive commutative groups to types creates all limits.",
+  to_additive_relevant_arg 2]
+noncomputable instance forget_createsLimitsOfSize :
+    CreatesLimitsOfSize.{w, v} (forget CommGrp.{u}) where
+      CreatesLimitsOfShape := inferInstance
+
+@[to_additive]
+noncomputable instance forget_createsLimits : CreatesLimits (forget Grp.{u}) :=
+  Grp.forget_createsLimitsOfSize.{u, u}
+
```
Comment (joelriou) [inline Mathlib/Algebra/Category/Grp/Limits.lean:456]: This is redundant with `Grp.forget_createsLimitsOfSize`:
```suggestion
```

File: Mathlib/Algebra/Category/Grp/Limits.lean [hunk 12]
```diff
@@ -193,6 +197,33 @@ instance forget_preservesLimitsOfSize :
...
+  "The forgetful functor from additive groups to types creates all limits.",
+  to_additive_relevant_arg 2]
+noncomputable instance forget_createsLimitsOfSize :
+    CreatesLimitsOfSize.{w,v} (forget Grp.{u}) where
+      CreatesLimitsOfShape := inferInstance
+
+@[to_additive]
+noncomputable instance forget_createsLimits :
+    CreatesLimits (forget Grp.{u}) := Grp.forget_createsLimitsOfSize.{u,u}
+
```
Comment (joelriou) [inline Mathlib/Algebra/Category/Grp/Limits.lean:226]: `CreatesLimits` is a particular case (abbrev) of `CreateLimitsOfSize`:
```suggestion
```

File: Mathlib/Algebra/Category/Grp/Limits.lean [hunk 13]
```diff
@@ -193,6 +197,33 @@ instance forget_preservesLimitsOfSize :
...
+    CreatesLimitsOfShape J (forget Grp.{u}) where
+      CreatesLimit := inferInstance
+
+/-- The forgetful functor from groups to types creates all limits.
+-/
+@[to_additive
+  "The forgetful functor from additive groups to types creates all limits.",
+  to_additive_relevant_arg 2]
+noncomputable instance forget_createsLimitsOfSize :
+    CreatesLimitsOfSize.{w,v} (forget Grp.{u}) where
```
Comment (joelriou) [inline Mathlib/Algebra/Category/Grp/Limits.lean:220]:
```suggestion
    CreatesLimitsOfSize.{w, v} (forget Grp.{u}) where
```

File: Mathlib/Algebra/Category/Grp/Limits.lean (hunk 2 above)
Comment (smorel394) [reply Mathlib/Algebra/Category/Grp/Limits.lean:93]: Done and done.

**Comment by joelriou** [comment]
Thanks!

bors merge

**Pull request closed without merging**